tests/test_preprints.py

```diff
@@ -794,3 +796,49 @@ def test_creator_gets_email(self, send_mail):
 
         self.preprint_branded.set_published(True, auth=Auth(self.user), save=True)
         assert_equals(send_mail.call_count, 2)
+
+class TestWithdrawnPreprint(OsfTestCase):
+    def setUp(self):
+        super(TestWithdrawnPreprint, self).setUp()
+        self.user = AuthUserFactory()
+        self.preprint_pre_mod = PreprintFactory(provider__reviews_workflow='pre-moderation', is_published=False)
+        self.preprint_post_mod = PreprintFactory(provider__reviews_workflow='post-moderation', is_published=False)
+        self.preprint = PreprintFactory()
+
+    @mock.patch('website.preprints.tasks.get_and_set_preprint_identifiers')
+    def test_withdrawn_preprint(self, _):
+        # test_ever_public
+
+        # non-moderated
+        assert_true(self.preprint.ever_public)
+
+        # pre-mod
+        self.preprint_pre_mod.run_submit(self.user)
+
+        assert_false(self.preprint_pre_mod.ever_public)
+        self.preprint_pre_mod.run_reject(self.user, 'it')
+        self.preprint_pre_mod.reload()
+        assert_false(self.preprint_pre_mod.ever_public)
+        self.preprint_pre_mod.run_accept(self.user, 'it')
+        self.preprint_pre_mod.reload()
+        assert_true(self.preprint_pre_mod.ever_public)
+
+        # post-mod
+        self.preprint_post_mod.run_submit(self.user)
+
+        assert_false(self.preprint_post_mod.ever_public)
+        self.preprint_post_mod.run_reject(self.user, 'it')
+        self.preprint_post_mod.reload()
+        assert_false(self.preprint_post_mod.ever_public)
+        self.preprint_post_mod.run_accept(self.user, 'it')
+        self.preprint_post_mod.reload()
+        assert_true(self.preprint_post_mod.ever_public)
+
+        # test_cannot_set_ever_public_to_False
+        self.preprint_pre_mod.ever_public = False
+        self.preprint_post_mod.ever_public = False
+        self.preprint.ever_public = False
+        with assert_raises(ValidationError):
```

Comment:
This will only tell you if any one of the things executed below raises a validation error. To ensure that each one does, you'll need to put each `.save` call in its own assert.

Additionally, this statement could be swapped out for `with pytest.raises(ValidationError):`. However, because `pytest` isn't already in scope here, this part isn't a blocker.